haystack/agents/base.py

```diff
@@ -0,0 +1,271 @@
+            "---\n\n"
+            "Question: $query\n"
+            "Thought: Let's think step-by-step, I first need to "
+        )
+
+    def add_tool(self, tool: Tool):
+        """
+        Add a new tool to the Agent and update the template for the Agent's PromptNode.
+
+        :param tool: The Tool to add to the Agent. Any previously added tool with the same name will be overwritten.
+        """
+        self.tools[tool.name] = tool
+        self.prompt_text = self._generate_prompt_text()
+
+    def run(self, query: str, params: Optional[dict] = None):
+        """
+        Runs the Agent given a query and optional parameters to pass on to the tools used. The result is in the
+        same format as a pipeline's result: a dictionary with a key "answers" containing a List of Answers
+
+        :param query: The search query (for query pipelines only).
+        :param params: A dictionary of parameters that you want to pass to those tools that are pipelines.
+                       To pass a parameter to all nodes in those pipelines, use: `{"top_k": 10}`.
+                       To pass a parameter to targeted nodes in those pipelines, use:
+                        `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        if len(self.tools) == 0:
+            raise AgentError("Agents without tools cannot be run. Add at least one tool using add_tool().")
+        if params and "max_iterations" in params:
+            remaining_iterations = params["max_iterations"]
+        else:
+            remaining_iterations = self.max_iterations
+
+        transcript = self.prompt_text
+        final_answer = ""
+
+        # In each iteration, choose a tool, generate input for it, and run it until the final answer is found or the
+        # maximum number of iterations is reached.
+        while remaining_iterations > 0:
+            remaining_iterations -= 1
+
+            prompt_template = PromptTemplate("think-step-by-step", transcript, ["query"])
+            pred = self.prompt_node.prompt(prompt_template=prompt_template, query=query)
+            transcript += f"{pred[0]}\n"
+
+            tool_name, tool_input = self._parse_tool_name_and_tool_input(pred=pred[0])
+            if tool_name == "Final Answer":
+                final_answer = tool_input
+                break
+            if tool_name not in self.tools:
+                # TODO we could try to recover from this error by rephrasing the question and retrying
+                raise AgentError(
+                    f'The Agent tried to use a tool "{tool_name}" but the registered tools are only {self.tools.keys()}.'
+                )
+            pipeline_or_node = self.tools[tool_name].pipeline_or_node
+            # We can only pass params to pipelines but not to nodes
```

Comment:
I would replace everything from 166 to 189 with method call `_extract_observation`, which has the parameter `tool_name`. Let's hide all this nitty-gritty logic from the main loop